docs/design/query-caching.md

```diff
@@ -0,0 +1,224 @@
+
+## Why cache?
+
+Reduce the number of round-trip requests between Data API builder (DAB) and the configured back-end database. Reducing the number of round-trip requests results in the end-user experiencing quicker response times.
+
+### Objectives
+
+- Cache responses to read operations
+    - REST: GET requests REST
+    - GraphQL: Queries
+- Make cache configuration simple.
+    - Limit developer exposure to implementation details of the cache and only surface essential properties in the runtime configuration.
+    - Runtime config properties remain consistent and optional for backwards and future compatibility.
+    - Replacing cache implementation (e.g. libraryA to libraryB) does not observably affect developers and their configuration.
+
+### Non-Objectives
+
+-	Follow-up reads on mutations (REST/GraphQL) are not cached.
+-	Distributed caching for syncing cache entries across servers for multiple instances of DAB.
+
+## Design
+
+The caching service should be database and endpoint agnostic. DAB will instantiate a singleton caching service that can be used within dependent services. In order to reduce coupling between DAB and a particular caching service implementation, the most generic caching options are publicly exposed in the runtime configuration: cache enablement and cache entry ttl. Additional options specific to the caching service will be explicitly set internally by DAB.
+
+Intentionally surfacing limited configuration options to the developer enables us to __make the caching service function uniformaly and independent of local hosting and SWA hosting configuration.__ This is because the developer exposed configuration properties are applied per-request and unnecessary for configuring the cache service during application startup.
+
+To reduce the sprawl of cache implementation across the code base, we will introduce the caching layer within __a common step of query execution for both REST and GraphQL: the query engine.__ The query engine's responsibility is to build a database query and dispatch the query to the appropriate query executor aligned to the backend database type. 
+
+By keeping the caching layer out of the query executor layer, we eliminate dependencies between database specific query execution logic and caching. Additionally, we prevent adding even tighter coupling between the runtime configuration and query executor because the query executor will not need to reference the runtime config to determine whether caching a request is necessary.
+
+### Cache entry
+
+#### Key
+
+`dataSourceName:queryText:queryParameters`
+
+The following example illustrates a real value created using the key schema listed above (new lines added for readability):
+
+```text
+7a07f92a-1aa2-4e2a-81d6-b9af0a25bbb6:
```

Comment:
in case of multiple data sources, this may result in potentially heterogenous query texts - belonging to potentially different types of backends within the same cache, e.g. sql query for mssql v/s pgsql.

How does this affect caching for Cosmos db sources?

Reply:
at the moment, Cosmos is not included since Cosmos uses a distinct `CosmosQueryEngine`. I know this adds drift in functionality between SQL dbs and Cosmos, but would need more Cosmos involvement to help implement with their query engine.

Reply:
Would you please create a separate issue for CosmosDb sources, make Sajee and Neeraj aware of the same?

Reply:
[#1884](https://github.com/Azure/data-api-builder/issues/1884)